File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/connectors/HttpConnector.java
package ru.kmorozov.gbd.core.logic.connectors;

import com.google.api.client.http.HttpHeaders;
import org.apache.commons.lang3.StringUtils;
import org.jsoup.nodes.Document;
import org.jsoup.parser.Parser;
import ru.kmorozov.db.utils.Mapper;
import ru.kmorozov.gbd.core.logic.Proxy.HttpHostExt;
import ru.kmorozov.gbd.core.logic.Proxy.UrlType;
import ru.kmorozov.gbd.logger.Logger;

import java.io.IOException;
import java.io.StringReader;
import java.nio.charset.Charset;
import java.util.Map;

/**
 * Created by km on 17.05.2016.
 */
public abstract class HttpConnector implements AutoCloseable {

    public static final int CONNECT_TIMEOUT = 30000;
    protected static final Logger logger = Logger.getLogger(HttpConnector.class);
    protected static final int MAX_RETRY_COUNT = 2;
    protected static final int SLEEP_TIME = 500;

    private static final Parser parser = Parser.htmlParser();

    protected static String getProxyKey(HttpHostExt proxy) {
        return proxy.toString();
    }

    public abstract Response getContent(String url, HttpHostExt proxy, boolean withTimeout) throws IOException;

    public Document getHtmlDocument(final String url, final HttpHostExt proxy, final boolean withTimeout) throws IOException {
        try {
            final Response response = this.getContent(url, proxy, withTimeout);
            if (response == null)
                throw new IOException("Cannot get document!");

            return HttpConnector.parser.parseInput(new StringReader(new String(response.getContent().readAllBytes(), Charset.forName("UTF-8"))), url);
        } finally {
            proxy.updateTimestamp();
        }
    }

    public Map<String, String> getJsonMapDocument(final String url, final HttpHostExt proxy, final boolean withTimeout) throws IOException {
        try {
            final Response response = this.getContent(url, proxy, withTimeout);
            if (response == null)
                throw new IOException("Cannot get document!");

            return Mapper.getGson().fromJson(new String(response.getContent().readAllBytes(), Charset.forName("UTF-8")), Mapper.mapType);
        } finally {
            proxy.updateTimestamp();
        }
    }

    public String getString(final String url, final HttpHostExt proxy, final boolean withTimeout) throws IOException {
        try {
            final Response response = this.getContent(url, proxy, withTimeout);
            if (response == null)
                throw new IOException("Cannot get document!");

            return new String(response.getContent().readAllBytes(), Charset.forName("UTF-8"));
        } finally {
            proxy.updateTimestamp();
        }
    }

    public UrlType getUrlType(final String url) {
        if (url.contains("books.google"))
            return UrlType.GOOGLE_BOOKS;
        else if (url.contains("jstor"))
            return UrlType.JSTOR;
        else
            return UrlType.OTHER;
    }

    protected boolean needHeaders(final String url) {
        return this.getUrlType(url) != UrlType.OTHER;
    }

    protected boolean validateProxy(final String url, final HttpHostExt proxy) {
        if (!this.needHeaders(url))
            return true;

        final UrlType urlType = this.getUrlType(url);

        final HttpHeaders headers = proxy.getHeaders(urlType);

        if (StringUtils.isEmpty(headers.getCookie()))
            return false;

        switch (urlType) {
            case GOOGLE_BOOKS:
                return headers.getCookie().contains("NID");
            case JSTOR:
                return headers.getCookie().contains("UUID");
            default:
                return true;
        }
    }
}

File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/connectors/HttpConnector.kt
package ru.kmorozov.gbd.core.logic.connectors

import com.google.common.base.Strings
import org.jsoup.nodes.Document
import org.jsoup.parser.Parser
import ru.kmorozov.db.utils.Mapper
import ru.kmorozov.gbd.core.config.GBDOptions
import ru.kmorozov.gbd.core.logic.proxy.HttpHostExt
import ru.kmorozov.gbd.core.logic.proxy.UrlType
import ru.kmorozov.gbd.logger.Logger
import java.io.IOException
import java.io.InputStreamReader
import java.nio.charset.Charset

/**
 * Created by km on 17.05.2016.
 */
abstract class HttpConnector : AutoCloseable {

    @Throws(IOException::class)
    abstract fun getContent(url: String, proxy: HttpHostExt, withTimeout: Boolean): Response

    @Throws(IOException::class)
    fun getHtmlDocument(url: String, proxy: HttpHostExt, withTimeout: Boolean): Document {
        try {
            val response = getContent(url, proxy, withTimeout)

            return parser.parseInput(InputStreamReader(response.content), url)
        } finally {
            proxy.updateTimestamp()
        }
    }

    @Throws(IOException::class)
    fun getJsonMapDocument(url: String, proxy: HttpHostExt, withTimeout: Boolean): Map<String, String> {
        try {
            val response = getContent(url, proxy, withTimeout)

            return Mapper.gson.fromJson(String(response.content.readAllBytes(), Charset.forName("UTF-8")), Mapper.mapType)
        } finally {
            proxy.updateTimestamp()
        }
    }

    @Throws(IOException::class)
    fun getString(url: String, proxy: HttpHostExt, withTimeout: Boolean): String {
        try {
            val response = getContent(url, proxy, withTimeout)

            return String(response.content.readAllBytes(), Charset.forName("UTF-8"))
        } finally {
            proxy.updateTimestamp()
        }
    }

    fun getUrlType(url: String): UrlType {
        return if (url.contains("frontcover"))
            UrlType.GOOGLE_BOOK_INFO
        else if (url.contains("books.google"))
            UrlType.GOOGLE_BOOKS
        else if (url.contains("jstor"))
            UrlType.JSTOR
        else
            UrlType.OTHER
    }

    protected fun needHeaders(url: String): Boolean {
        return getUrlType(url) == UrlType.GOOGLE_BOOKS || getUrlType(url) == UrlType.GOOGLE_BOOK_INFO
    }

    protected fun validateProxy(url: String, proxy: HttpHostExt): Boolean {
        if (!needHeaders(url))
            return true

        val urlType = getUrlType(url)

        val headers = proxy.getHeaders(urlType)

        if (Strings.isNullOrEmpty(headers.cookie))
            return false

        return when (urlType) {
            UrlType.GOOGLE_BOOKS -> headers.cookie.contains("NID")
            UrlType.JSTOR -> headers.cookie.contains("UUID")
            else -> true
        }
    }


    protected fun getProxyKey(proxy: HttpHostExt): String {
        return proxy.toString()
    }

    companion object {

        const val CONNECT_TIMEOUT = 30000
        private val logger = Logger.getLogger(GBDOptions.debugEnabled, HttpConnector::class.java)
        const val MAX_RETRY_COUNT = 5
        const val SLEEP_TIME = 500

        private val parser = Parser.htmlParser()
    }
}
